processor/schemaprocessor/internal/translation/translation.go

```diff
@@ -0,0 +1,341 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package translation // import "github.com/open-telemetry/opentelemetry-collector-contrib/processor/schemaprocessor/internal/translation"
+
+import (
+	"context"
+	"io"
+	"sort"
+
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	encoder "go.opentelemetry.io/otel/schema/v1.0"
+	"go.opentelemetry.io/otel/schema/v1.0/ast"
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/processor/schemaprocessor/internal/alias"
+)
+
+// Translation defines the complete abstraction of schema translation file
+// that is defined as part of the https://opentelemetry.io/docs/reference/specification/schemas/file_format_v1.0.0/
+// Each instance of Translation is "Target Aware", meaning that given a schemaURL as an input
+// it will convert from the given input, to the configured target.
+//
+// Note: as an optimisation, once a Translation is returned from the manager,
+//
+//	there is no checking the incoming signals if the schema family is a match.
+type Translation interface {
+	// SupportedVersions checks to see if the provided version is defined as part
+	// of this translation since it is useful to know it the translation is missing
+	// updates.
+	SupportedVersion(v *Version) bool
+
+	// ApplyAllResourceChanges will modify the resource part of the incoming signals
+	// This applies to all telemetry types and should be applied there
+	ApplyAllResourceChanges(ctx context.Context, in alias.Resource, inSchemaURL string) error
```

Comment:
`alias.Resource` already has a SchemaURL() getter. Why is `inSchemaURL` needed here? Perhaps use `pdata.Resource` instead for symmetry with the rest of methods in this interface?

Reply:
Ok I figured it out - `ApplyAllResourceChanges` is responsible for changing the SchemaURL of each `ResourceX`.  Similarly, `ApplyScopeXChanges` changes the SchemaURL of each `ScopeX`.  Not sure what I'm going to do here because I do like the idea of just using `pcommon.Resource` instead of the top level `ResourceX` in `ApplyAllResourceChanges`

Reply:
I think I'm going to keep it as is for now... I think keeping the symmetry is worth the slight confusion for `ApplyAllResourceChanges`.  Similarly, keeping, `alias.Resource` is here since all the other fns are responsible for modifying the SchemaURL.  Open to changing the name of `alias.Resource` if you have ideas..